File: solutions/22ContainsDuplicateIII/java/ContainsDuplicateIII.java
import java.util.HashMap;

class ContainsDuplicateIII {
  public boolean containsNearbyAlmostDuplicate(int[] nums, int indexDiff, int valueDiff) {
    TreeSet<Long> set = new TreeSet<>();
    for (int i = 0; i < nums.length; i++) {
      Long floor = set.floor((long) nums[i] + valueDiff);
      Long ceiling = set.ceiling((long) nums[i] - valueDiff);
      if ((floor != null && floor >= nums[i]) || (ceiling != null && ceiling <= nums[i]))
        return true;
      set.add((long) nums[i]);
      if (i >= indexDiff)
        set.remove((long) nums[i - indexDiff]);
    }
    return false;
  }

}

File: solutions/22ContainsDuplicateIII/python/ContainsDuplicateIII.py
from typing import List


class Solution(object):
    def containsNearbyAlmostDuplicate(
        self, nums: List[int], indexDiff: int, valueDiff: int
    ) -> bool:
        # n = len(nums)
        # for i in range(n):
        #     for j in range(i + 1, min(i + indexDiff + 1, n)):
        #         if abs(nums[i] - nums[j]) <= valueDiff:
        #             return True
        # return False
        bucket_dict = dict()
        for i in range(len(nums)):
            num = nums[i] // (valueDiff + 1)
            if num in bucket_dict:
                return True
            bucket_dict[num] = nums[i]

            if (num - 1) in bucket_dict and abs(
                bucket_dict[num - 1] - nums[i]
            ) <= valueDiff:
                return True
            if (num + 1) in bucket_dict and abs(
                bucket_dict[num + 1] - nums[i]
            ) <= valueDiff:
                return True
            if i >= indexDiff:
                bucket_dict.pop(nums[i - indexDiff] // (valueDiff + 1))
        return False
